sw/device/silicon_creator/rom/e2e/rom_e2e_static_critical_test.c

```diff
@@ -0,0 +1,142 @@
+  CHECK((uintptr_t)addr_ % sizeof(uint32_t) == 0, \
+        #addr_ " must be word aligned");
+
+void boot_measurements_test(void) {
+  CHECK(kManifest.usage_constraints.selector_bits == 0,
+        "Selector bits must be 0");
+  const volatile char *manifest_start = (const volatile char *)&kManifest;
+  const char *manifest_end = (const char *)manifest_start + sizeof(manifest_t);
+  const volatile char *signed_region_start =
+      manifest_start + sizeof(sigverify_rsa_buffer_t);
+  const char *signed_region_end =
+      (const char *)manifest_start + kManifest.length;
+  size_t manifest_signed_region_size = manifest_end - signed_region_start;
+  size_t signed_region_size = signed_region_end - signed_region_start;
+  dif_hmac_t hmac;
+
+  CHECK_WORD_ALIGNED(manifest_start);
+  CHECK_WORD_ALIGNED(manifest_end);
+  CHECK_WORD_ALIGNED(signed_region_start);
+  CHECK_WORD_ALIGNED(signed_region_end);
+  CHECK_WORD_ALIGNED(manifest_signed_region_size);
+  CHECK_WORD_ALIGNED(signed_region_size);
+
+  LOG_INFO("Manifest: [0x%p, 0x%p)", manifest_start, manifest_end);
+  LOG_INFO("ROM_EXT signed region: [0x%p, 0x%p)", signed_region_start,
+           signed_region_end);
+  CHECK_DIF_OK(
+      dif_hmac_init(mmio_region_from_addr(TOP_EARLGREY_HMAC_BASE_ADDR), &hmac));
+  CHECK_DIF_OK(dif_hmac_mode_sha256_start(
+      &hmac, (dif_hmac_transaction_t){
+                 .digest_endianness = kDifHmacEndiannessLittle,
+                 .message_endianness = kDifHmacEndiannessLittle,
+             }));
+
+  // Copy the part of the manifest that's in the signed region to
+  // memory before pushing to hmac since it's volatile.
+  // Note: this array is larger than `manifest_signed_region_size` since VLAs
+  // are optional in C11.
+  char manifest_signed_region[sizeof(manifest_t)];
+  for (size_t i = 0; i < manifest_signed_region_size; ++i) {
+    manifest_signed_region[i] =
+        *((const volatile char *)signed_region_start + i);
+  }
+  hmac_testutils_push_message(&hmac, manifest_signed_region,
+                              manifest_signed_region_size);
+  // Rest of the image
+  hmac_testutils_push_message(&hmac, manifest_end,
+                              signed_region_size - manifest_signed_region_size);
+
+  CHECK_DIF_OK(dif_hmac_process(&hmac));
+  dif_hmac_digest_t act_digest;
+  hmac_testutils_finish_polled(&hmac, &act_digest);
+
+  LOG_INFO("SHA256(static_critical.boot_measurements): 0x%!x",
+           sizeof(boot_measurements), &boot_measurements);
+  LOG_INFO("SHA256(actual digest): 0x%!x", sizeof(act_digest), &act_digest);
+  CHECK_ARRAYS_EQ(boot_measurements.rom_ext.data, act_digest.digest,
+                  ARRAYSIZE(act_digest.digest));
+}
+
+void sec_mmio_pos_test(void) {
+  enum {
+    kRndOffset = 13,
+    kRomCheckValue = 6,
+  };
+
+  sec_mmio_check_values(kRndOffset);
+  sec_mmio_check_counters(kRomCheckValue + 1);
+}
+
+static bool exception_expected = false;
+static bool exception_observed = false;
```

Comment:
nit: just these be volatile because `ottr_exeception_handler` is accessing them? Or is that now old school :)

Reply:
Thanks for catching this, yes these should be `volatile`.